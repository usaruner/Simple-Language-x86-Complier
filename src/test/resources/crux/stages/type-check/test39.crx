// test arithmetic

func test(a:int, b:int) : void {
  ::printInt(a + b);
  ::println();
  ::printInt(a - b);
  ::println();
  ::printInt(a * b);
  ::println();
  if (b != 0) {
    ::printInt(a / b);
    ::println();
  } else {
    ::println();
  }
}

func main() : void {
  var i:int;
  var j:int;
  let i = 0;
  while(i <= 10) {
    let j = 0;
    while(j <= 10) {
       ::test(i, j);
       let j = j + 1;
    }
    let i = i + 1;
  }
}
