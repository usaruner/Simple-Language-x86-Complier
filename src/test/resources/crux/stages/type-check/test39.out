Crux Program has no type errors.
ast.DeclarationList(3)
  ast.FunctionDefinition(3)[Symbol(test:func(TypeList(int, int)):void), [Symbol(a:int), Symbol(b:int)]]
    ast.StatementList(4)
      void: ast.Call(4)[Symbol(printInt)]
        int: ast.OpExpr(4)[ADD]
          int: ast.Dereference(4)
            Address(int): ast.Name(4)[a]
          int: ast.Dereference(4)
            Address(int): ast.Name(4)[b]
      void: ast.Call(5)[Symbol(println)]
      void: ast.Call(6)[Symbol(printInt)]
        int: ast.OpExpr(6)[SUB]
          int: ast.Dereference(6)
            Address(int): ast.Name(6)[a]
          int: ast.Dereference(6)
            Address(int): ast.Name(6)[b]
      void: ast.Call(7)[Symbol(println)]
      void: ast.Call(8)[Symbol(printInt)]
        int: ast.OpExpr(8)[MULT]
          int: ast.Dereference(8)
            Address(int): ast.Name(8)[a]
          int: ast.Dereference(8)
            Address(int): ast.Name(8)[b]
      void: ast.Call(9)[Symbol(println)]
      ast.IfElseBranch(10)
        bool: ast.OpExpr(10)[NE]
          int: ast.Dereference(10)
            Address(int): ast.Name(10)[b]
          int: ast.LiteralInt(10)[0]
        ast.StatementList(11)
          void: ast.Call(11)[Symbol(printInt)]
            int: ast.OpExpr(11)[DIV]
              int: ast.Dereference(11)
                Address(int): ast.Name(11)[a]
              int: ast.Dereference(11)
                Address(int): ast.Name(11)[b]
          void: ast.Call(12)[Symbol(println)]
        ast.StatementList(14)
          void: ast.Call(14)[Symbol(println)]
  ast.FunctionDefinition(18)[Symbol(main:func(TypeList()):void), []]
    ast.StatementList(19)
      ast.VariableDeclaration(19)[Symbol(i:int)]
      ast.VariableDeclaration(20)[Symbol(j:int)]
      void: ast.Assignment(21)
        Address(int): ast.Name(21)[i]
        int: ast.LiteralInt(21)[0]
      ast.WhileLoop(22)
        bool: ast.OpExpr(22)[LE]
          int: ast.Dereference(22)
            Address(int): ast.Name(22)[i]
          int: ast.LiteralInt(22)[10]
        ast.StatementList(23)
          void: ast.Assignment(23)
            Address(int): ast.Name(23)[j]
            int: ast.LiteralInt(23)[0]
          ast.WhileLoop(24)
            bool: ast.OpExpr(24)[LE]
              int: ast.Dereference(24)
                Address(int): ast.Name(24)[j]
              int: ast.LiteralInt(24)[10]
            ast.StatementList(25)
              void: ast.Call(25)[Symbol(test)]
                int: ast.Dereference(25)
                  Address(int): ast.Name(25)[i]
                int: ast.Dereference(25)
                  Address(int): ast.Name(25)[j]
              void: ast.Assignment(26)
                Address(int): ast.Name(26)[j]
                int: ast.OpExpr(26)[ADD]
                  int: ast.Dereference(26)
                    Address(int): ast.Name(26)[j]
                  int: ast.LiteralInt(26)[1]
          void: ast.Assignment(28)
            Address(int): ast.Name(28)[i]
            int: ast.OpExpr(28)[ADD]
              int: ast.Dereference(28)
                Address(int): ast.Name(28)[i]
              int: ast.LiteralInt(28)[1]
